### Introduction

This project performs an analysis on data obtained from Amazon. Amazon, one of the largest e-commerce websites in the world, also runs fulfillment centres where they store and ship the inventory from. A better understanding of their customer base, such as the effects of discounts on sales or the effects of good rating scores on sales, can lead to increased efficiencies and reduced costs.

The dataset consists of reviews of products sold on Amazon. Examples of the data within the dataset are the ratings of the products, the names of the people who bought those products, and the discounts applied to the products.

The dataset was obtained from Kaggle:

https://www.kaggle.com/datasets/karkavelrajaj/amazon-sales-dataset

In [1]:
import pandas as pd 
import numpy as np 
import seaborn as sns
import plotly.express as px
import plotly.subplots as sp
import plotly.graph_objects as go
import warnings
import pygwalker as pyg
import sqlite3
#import sqlalchemy
#from sqlalchemy import create_engine
from sqlalchemy.dialects import sqlite
from pandas.io import sql
import subprocess


warnings.filterwarnings('ignore')

The dataset is in the **CSV** (Comma Seperated Value) format, and will be imported into an **SQLite** database, where it will be preprocessed.

In [2]:
df=pd.read_csv('amazon.csv')
conn=sqlite3.connect('amazon.db')
df.to_sql(name='amazon', con=conn, if_exists='replace', index=False)
conn.commit()

In [3]:
%config SqlMagic.displaylimit = 15

In [4]:
%load_ext sql
%sql sqlite:///amazon.db

We shall check to make sure that the CSV file has been imported correctly into the SQLite database and that we have managed to successfully connecct to it.

In [5]:
%sql select * from amazon limit 1

Running query in 'sqlite:///amazon.db'

+------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------+------------------+--------------+---------------------+--------+--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| product_id |                                                                            product_name                                                                            |                                      category                                     | discounted_price | actual_price | discount_percentage | rating | rating_count |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          about_product                                                                                                                                                                                                                                                                                                                                             

Next, we shall check to see if there are any duplicate reviews. We will not be using the user_name or user_id for this as the users could have bought multiple products. Luckily, each review will have its own unique review_id.

In [6]:
%%sql
SELECT review_id, COUNT(review_id) AS count, SUM(COUNT(review_id)) OVER() AS total_count
FROM amazon GROUP BY review_id having count(review_id)>1

Running query in 'sqlite:///amazon.db'

+-------------------------------------------------------------------------------------------------------------------------+-------+-------------+
|                                                        review_id                                                        | count | total_count |
+-------------------------------------------------------------------------------------------------------------------------+-------+-------------+
|    R10365HEDURWI9,R5RP542IMC4OI,RX2HFWXTTQDTS,R2636VYPMOZV9,RW2Z2YM3K8UV5,RVNGA0FEAXYHI,R2K7MABWMAQE26,R33YS4PO3JWU23   |   3   |     415     |
|  R10FUJSCR3VYHY,R2Y8B5LQ5HLACQ,R3BC8GS9GGMBTI,R2BO0XUUDY4ZA3,RN23FCU4EP3F3,RDGNXFM923PG4,R26PGAI8JKY8XB,R381CGOL80J2QM  |   2   |     415     |
| R10I6UIAQIP9TN,R2XEWWLV1LH7KX,R3J0MEY15WI71Z,R3HJ0GBBBUGEJZ,R3TGTIJ54KHOL0,R21TUQZLYNGC0M,R1JSFOA0TD4S1A,R1KOD8YMT3FJ7I |   2   |     415     |
|   R10KEMT1N336ZD,RL01KZO95GX4F,R1Q721FI3A7XLK,R34MTIAB8IHAI,R1LG1DNA516T7L,RFH8DR3A2O8BG,RFA922H587JFN,R10BFD806POSOX   |   2   |     415     |
|  R11MQS7WD9C3I0,R2AKH69XQY8BY4,R8GBOLYUN5UP6,R1AYVO4R25KJTA,R1HT6XM787V7FV,R339XJL1GMKHA3,R175VFSB2A32HG,R35T9LXYBSP09G |   3   |     415     |
|   R128LZ0DN2NZBZ,R3LFQ7EDHZ6DKM,RUSJFUV64DPWM,RHNVN7WEES6ZV,R3LHNY1FJU5Z62,RYD25TMDIWVXF,R22G4CIX0JF8CT,R3KZ4E667WBY58  |   3   |     415     |
|  R12D1BZF9MU8TN,R32MNCWO5LGFCG,RZU3UK8OZKD6X,R3BSTKR3JUW6GY,R1ARVYPXS4XPB7,R1V6GDYE2IBX8O,R28EG2PXZTJL90,R2SQNU7OIOOLHT |   3   |     415     |
|  R13UTIA6KOF6QV,R2UGDZSGFF01K7,RHHIZ45VYU5X6,R14N9HBE5EIUY0,R2WMW096T9Y0OU,R1SHIIE6M72825,R22P6BE9DBME4F,R2TEINENXTIHT2 |   5   |     415     |
|  R14ZOPYFHOYYRQ,R1GQH74NUCJZZ7,R1BNWIYBRSI1Z6,R347KU67LE6JEH,RMGA8IGV2WQDX,R2782FIPC5T4KM,R220M468LVHIE1,RA1PNAU355MLG  |   2   |     415     |
|                               R18D9LZAYX9JSY,R2TD56H4WD69RD,R3022ERQVPT7PV,R3T0CWF358RZNJ                               |   2   |     415     |
|   R19Q6OQ19PWL5K,RXWY3WK7QVN25,R10S2P5H6YODNY,R2ILGDHXO6XX4K,R2TWCN72P6DU1Y,ROTBOX5J8LVNW,R4PXSKQEZNJGO,R2DDR8ZR4YXV8M  |   2   |     415     |
| R1A2H4LNTTSZKN,R29RZ6S6SY3H4F,R2MZ7BZ4991B7O,R125UHW97PT3OH,R1GNNZDXKP43DG,R1ZDKQ5659C68H,R36FYJ9DGL1QL1,R1IZDBZW18XJPH |   3   |     415     |
|  R1AIQQLE21YDXS,R26ABOIUJ8UXJ7,R93L2MCBC4Y90,R2GDAM50Z413JN,R16TI1N60Q41BB,R1UEYEMD03OA5C,R16D88E4TNGL3M,R1WSNRYZ7VK0KB |   2   |     415     |
|  R1BC08IFG4REKS,R1FJKIHIO54SOW,R3JR48W2CI480,R3JH7SHSXDT1GT,R35QWAY83WL8H6,R25N2U90N2A5AS,R19AK3DT3JOE82,R210WJI15JCSRE |   3   |     415     |
|  R1BFOK13WV2QLM,R3H97FN1H50F7F,R1IY2IDRUJX5O5,R1N5UJPJ5YGBU5,R3BZ3W2KH0X1DQ,R3GPO2WYK6ABG,RCMFGYS1T27LL,R1D5OAMYO4526T  |   4   |     415     |
+-------------------------------------------------------------------------------------------------------------------------+-------+-------------+
Truncated to displaylimit of 15.

There are a total of 415 duplicated rows, and we shall remove the duplicated entries from the table.

In [7]:
%%sql

DELETE FROM amazon 
WHERE rowid > (
  SELECT MIN(rowid) FROM amazon a  
  WHERE a.review_id = amazon.review_id
);


Running query in 'sqlite:///amazon.db'

271 rows affected.

++
||
++
++

All duplicated entries were removed, leaving only one of the entries behind for each duplicated entries. A total of 271 rows were removed. To confirm that the SQL query has worked as intended, we shall check for any duplicated entries remaining.

In [8]:
%%sql
select review_id, count(review_id) from amazon group by (review_id) having count(review_id)>1

Running query in 'sqlite:///amazon.db'

+-----------+------------------+
| review_id | count(review_id) |
+-----------+------------------+
+-----------+------------------+

We will also check for null values in the dataset. 

In [9]:
%%sql
SELECT *
FROM amazon
WHERE (product_id or product_name or category or discounted_price or actual_price or discount_percentage or rating or rating_count or user_id or user_name or review_id) IS NULL

Running query in 'sqlite:///amazon.db'

+------------+--------------+----------+------------------+--------------+---------------------+--------+--------------+---------------+---------+-----------+-----------+--------------+----------------+----------+--------------+
| product_id | product_name | category | discounted_price | actual_price | discount_percentage | rating | rating_count | about_product | user_id | user_name | review_id | review_title | review_content | img_link | product_link |
+------------+--------------+----------+------------------+--------------+---------------------+--------+--------------+---------------+---------+-----------+-----------+--------------+----------------+----------+--------------+
+------------+--------------+----------+------------------+--------------+---------------------+--------+--------------+---------------+---------+-----------+-----------+--------------+----------------+----------+--------------+

After the preprocessing is done, we shall convert the SQL table to a Pandas dataframe, and pygwalker will be used to visualise the data and perform Exploratory Data Analysis. The SQLite database connection will be closed to free up resources.

In [38]:
df = pd.read_sql_query("SELECT * FROM amazon", conn)
#conn.commit()
#conn.close()

Let's do a quick check of the dataframe.

In [30]:
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [12]:
walker = pyg.walk(df)

Box(children=(HTML(value='<div id="ifr-pyg-0" style="height: auto">\n    <head>\n        <meta http-equiv="Con…

It will be very helpful to further breakdown the item categories so that we can group them within broader categories.

In [39]:
df['cat']=df['category'].str.split("|", n = -1, expand = False).str[0]
df['cat1']=df['category'].str.split("|", n = -1, expand = False).str[1]
df['cat2']=df['category'].str.split("|", n = -1, expand = False).str[2]
df['cat3']=df['category'].str.split("|", n = -1, expand = False).str[3]
df['cat4']=df['category'].str.split("|", n = -1, expand = False).str[4]
df['cat5']=df['category'].str.split("|", n = -1, expand = False).str[5]
df['cat6']=df['category'].str.split("|", n = -1, expand = False).str[6]
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  ...  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...  ...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...  ...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...  ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...  ...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...  ...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link                    cat  \
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  Computers&Accessories   
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  Computers&Accessories   
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  Computers&Accessories   
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  Computers&Accessories   
4  https://www.amazon.in/Portronics-Konnect-POR-1...  Computers&Accessories   

                      cat1                cat2    cat3       cat4 cat5 cat6  
0  Accessories&Peripherals  Cables&Accessories  Cables  USBCables  NaN  NaN  
1  Accessories&Peripherals  Cables&Accessories  Cables  USBCables  NaN  NaN  
2  Accessories&Peripherals  Cables&Accessories  Cables  USBCables  NaN  NaN  
3  Accessories&Peripherals  Cables&Accessories  Cables  USBCables  NaN  NaN  
4  Accessories&Peripherals  Cables&Accessories  Cables  USBCables  NaN  NaN  

[5 rows x 23 columns]